### PR TITLE
Implement retry logic

### PR DESCRIPTION
**Pull request by hposca**
This is based on the [recommendation made by Yori](https://github.com/gruntwork-io/cloud-nuke/pull/181#discussion_r594453058)

Currently this is useful when we are trying to delete a login profile that has been created recently, but can be used on other parts too.

~Based on what we have [on terratest](https://github.com/gruntwork-io/terratest/blob/master/modules/retry/retry.go). And adapted the tests to what we have here.~
Using [current latest `go-commons` release](https://github.com/gruntwork-io/go-commons/releases/tag/v0.8.2) to implement the retry logic

~NOTE: Depends on #181~

~Also, by [scout rule](https://www.matheus.ro/2017/12/11/clean-code-boy-scout-rule/), I removed the last dependency for `terratest` on the code.~ -> Remove `terratest` dependencies will appear in another PR, maybe

### CODE REVIEW

File: retry/retry.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
+package retry
```
Comment (yorinasub17) [inline retry/retry.go:1]: We have a version of this already in `kubergrunt`, so I extracted it out to `go-commons` so we can use it here: https://github.com/gruntwork-io/go-commons/pull/41

**Review by yorinasub17 — COMMENTED** [review]
LGTM! However, I'd like to DRY the `retry` function further. Let's get `go-commons` port merged, and then update this to use that.

File: retry/retry.go (hunk 1 above)
Comment (yorinasub17) [reply retry/retry.go:1]: FYI the `go-commons` retry function is now merged and available! https://github.com/gruntwork-io/go-commons/releases/tag/v0.8.2

File: aws/ecs_utils_for_test.go [hunk 2]
```diff
@@ -15,11 +15,11 @@ import (
 	"github.com/aws/aws-sdk-go/service/iam"
 	"github.com/gruntwork-io/gruntwork-cli/collections"
 	gruntworkerrors "github.com/gruntwork-io/gruntwork-cli/errors"
-	"github.com/gruntwork-io/terratest/modules/retry"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 
 	"github.com/gruntwork-io/cloud-nuke/logging"
+	"github.com/gruntwork-io/cloud-nuke/retry"
```
Comment (yorinasub17) [inline aws/ecs_utils_for_test.go:22]: Oh this should actually be using the terratest `retry` function, which is optimized for testing (that is, it will fail the test if there is an error, and will log messages in a way that is friendly to `terratest_log_parser`).

**Comment by hposca** [comment]
To update the code to use the newest version of `go-commons` I ended up doing a bit more than I was expecting and want some feedback to check if this process is expected as this ended up being more a golang discovery journey...

To use `go-commons` I had to do a `go get github.com/gruntwork-io/go-commons`, which then updated `go.mod` and `go.sum` to use newer versions of some packages, specially the `aws-sdk-go` from `v1.22.3` to `v1.34.27`. This [broke the tests](https://app.circleci.com/pipelines/github/gruntwork-io/cloud-nuke/3123/workflows/51c965ff-1b07-4f7f-ab3e-540f4a165114/jobs/23768) because the underlying API on the SDK changed, which broke the Mocks. Then I had to run `mockgen` to update `mocks/EC2API.go` and then the tests went green.

❓Is this normal in the Golang world?❓

- There is no problem in the "to use this library you need to update other libraries" situation but at the same time,
- It doesn't feel right to be force-pushed to use a newer version of a super important library as the `aws-sdk`
  - In my mind this entire forced update to use the newest `aws-sdk` doesn't make too much sense as [`go-commons`](https://github.com/gruntwork-io/go-commons/blob/master/go.mod) doesn't even use `aws-sdk-go` 🤔
  - Googling for "how to update a single package" and similar queries didn't return anything useful (It seems that this is the expected golang behaviour)
    - From the Ruby world, updating a gem usually updates only the directly required gems

**Comment by zackproser** [comment]
> To update the code to use the newest version of `go-commons` I ended up doing a bit more than I was expecting and want some feedback to check if this process is expected as this ended up being more a golang discovery journey...
> 
> To use `go-commons` I had to do a `go get github.com/gruntwork-io/go-commons`, which then updated `go.mod` and `go.sum` to use newer versions of some packages, specially the `aws-sdk-go` from `v1.22.3` to `v1.34.27`. This [broke the tests](https://app.circleci.com/pipelines/github/gruntwork-io/cloud-nuke/3123/workflows/51c965ff-1b07-4f7f-ab3e-540f4a165114/jobs/23768) because the underlying API on the SDK changed, which broke the Mocks. Then I had to run `mockgen` to update `mocks/EC2API.go` and then the tests went green.
> 
> questionIs this normal in the Golang world?question
> 
>     * There is no problem in the "to use this library you need to update other libraries" situation but at the same time,
> 
>     * It doesn't feel right to be force-pushed to use a newer version of a super important library as the `aws-sdk`
>       
>       * In my mind this entire forced update to use the newest `aws-sdk` doesn't make too much sense as [`go-commons`](https://github.com/gruntwork-io/go-commons/blob/master/go.mod) doesn't even use `aws-sdk-go` thinking
>       * Googling for "how to update a single package" and similar queries didn't return anything useful (It seems that this is the expected golang behaviour)
>         
>         * From the Ruby world, updating a gem usually updates only the directly required gems

Oof, that is painful. I have not personally had issues where a particular version of a dependent package ended up causing issues, but have had lots of issues with `goimports -w` which I run on save and go modules fighting each other especially when there's a been a hard fork of a particular package as was the case with `go-git`(https://github.com/go-git/go-git#project-status) which was super fun resolve - as all the structs and methods were identical - so `goimports` kept swapping in the wrong one. 

I'm not sure if this answer on pinning particular modules https://stackoverflow.com/questions/53682247/how-to-point-go-module-dependency-in-go-mod-to-a-latest-commit-in-a-repo would have helped you in this case - that being said, if you were able to update the mocks and ended up upgrading to a newer package version :man_shrugging: In other instances, `go mod tidy` [can sometimes be of assistance](https://golangbyexample.com/go-mod-tidy/)

@brikis98 @bwhaley and @yorinasub17 will have more insight into the particulars of these tests.

**Comment by rhoboat** [comment]
I've seen this kind of issue before when updating go lang modules. When you have transitive dependencies which change, i.e., your repo depends on -> A which depends on -> B, and B has changed. See [the blocker in terragrunt for updating terraform to 0.14 internally](https://github.com/gruntwork-io/terragrunt/issues/1505).

**Comment by hposca** [comment]
I was finally able to understand why those things happened 😄

First, thank you all for the comments and links they were the starting point for the discovery 😄

TL;DR: It seems that this is expected as `github.com/gruntwork-io/go-commons@v0.8.2` depends on `github.com/gruntwork-io/kubergrunt@v0.6.10` which depends on `github.com/aws/aws-sdk-go@v1.34.27`

**TL version**

Spiralling through Google I found out [this SO answer](https://stackoverflow.com/a/63074019) that pointed out the existence of `go mod graph` which prints the module requirement graph (for each line: before the whitespace is the module and after it is one of its dependencies). So by doing a reverse search, beginning with `aws-sdk-go@v1.34.27`:

```
go mod graph | grep aws-sdk-go@v1.34.27

github.com/gruntwork-io/cloud-nuke github.com/aws/aws-sdk-go@v1.34.27
github.com/gruntwork-io/kubergrunt@v0.6.10 github.com/aws/aws-sdk-go@v1.34.27
github.com/aws/aws-sdk-go@v1.34.27 golang.org/x/net@v0.0.0-20200202094626-16171245cfb2
github.com/aws/aws-sdk-go@v1.34.27 github.com/pkg/errors@v0.9.1
github.com/aws/aws-sdk-go@v1.34.27 github.com/jmespath/go-jmespath@v0.3.0
github.com/aws/aws-sdk-go@v1.34.27 github.com/go-sql-driver/mysql@v1.5.0
```

Looking at the `github.com/gruntwork-io/kubergrunt@v0.6.10 github.com/aws/aws-sdk-go@v1.34.27` line it says that `kubergrunt` depends on this `aws-sdk-go` version... Then, trying to find who depends on `kubergrunt`:

```
go mod graph | grep github.com/gruntwork-io/kubergrunt@v0.6.10

github.com/gruntwork-io/kubergrunt@v0.6.10 sigs.k8s.io/aws-iam-authenticator@v0.5.1
github.com/gruntwork-io/kubergrunt@v0.6.10 k8s.io/client-go@v0.18.3
github.com/gruntwork-io/kubergrunt@v0.6.10 k8s.io/apimachinery@v0.18.3
github.com/gruntwork-io/kubergrunt@v0.6.10 k8s.io/api@v0.18.3
github.com/gruntwork-io/kubergrunt@v0.6.10 github.com/urfave/cli@v1.22.4
github.com/gruntwork-io/kubergrunt@v0.6.10 github.com/stretchr/testify@v1.6.1
github.com/gruntwork-io/kubergrunt@v0.6.10 github.com/sirupsen/logrus@v1.6.0
github.com/gruntwork-io/kubergrunt@v0.6.10 github.com/mitchellh/go-homedir@v1.1.0
github.com/gruntwork-io/kubergrunt@v0.6.10 github.com/gruntwork-io/terratest@v0.30.0
github.com/gruntwork-io/kubergrunt@v0.6.10 github.com/gruntwork-io/gruntwork-cli@v0.7.0
github.com/gruntwork-io/kubergrunt@v0.6.10 github.com/blang/semver/v4@v4.0.0
github.com/gruntwork-io/kubergrunt@v0.6.10 github.com/aws/aws-sdk-go@v1.34.27
github.com/gruntwork-io/go-commons@v0.8.2 github.com/gruntwork-io/kubergrunt@v0.6.10
```

This line `github.com/gruntwork-io/go-commons@v0.8.2 github.com/gruntwork-io/kubergrunt@v0.6.10` tells that it is required by `go-commons` 😄

So, in summary, `github.com/gruntwork-io/go-commons@v0.8.2` depends on `github.com/gruntwork-io/kubergrunt@v0.6.10` which depends on `github.com/aws/aws-sdk-go@v1.34.27` which then caused this required update 😄

I also found on the [go mod documentation](https://github.com/golang/go/wiki/Modules#why-does-go-mod-tidy-record-indirect-and-test-dependencies-in-my-gomod) that there is the `go mod why` command but it wasn't helpful 😞

**Review by zackproser — APPROVED** [review]
LGTM

**Comment by zackproser** [comment]
> I was finally able to understand why those things happened smile
> 
> First, thank you all for the comments and links they were the starting point for the discovery smile
> 
> TL;DR: It seems that this is expected as `github.com/gruntwork-io/go-commons@v0.8.2` depends on `github.com/gruntwork-io/kubergrunt@v0.6.10` which depends on `github.com/aws/aws-sdk-go@v1.34.27`
> 
> **TL version**
> 
> Spiralling through Google I found out [this SO answer](https://stackoverflow.com/a/63074019) that pointed out the existence of `go mod graph` which prints the module requirement graph (for each line: before the whitespace is the module and after it is one of its dependencies). So by doing a reverse search, beginning with `aws-sdk-go@v1.34.27`:
> 
> ```
> go mod graph | grep aws-sdk-go@v1.34.27
> 
> github.com/gruntwork-io/cloud-nuke github.com/aws/aws-sdk-go@v1.34.27
> github.com/gruntwork-io/kubergrunt@v0.6.10 github.com/aws/aws-sdk-go@v1.34.27
> github.com/aws/aws-sdk-go@v1.34.27 golang.org/x/net@v0.0.0-20200202094626-16171245cfb2
> github.com/aws/aws-sdk-go@v1.34.27 github.com/pkg/errors@v0.9.1
> github.com/aws/aws-sdk-go@v1.34.27 github.com/jmespath/go-jmespath@v0.3.0
> github.com/aws/aws-sdk-go@v1.34.27 github.com/go-sql-driver/mysql@v1.5.0
> ```
> 
> Looking at the `github.com/gruntwork-io/kubergrunt@v0.6.10 github.com/aws/aws-sdk-go@v1.34.27` line it says that `kubergrunt` depends on this `aws-sdk-go` version... Then, trying to find who depends on `kubergrunt`:
> 
> ```
> go mod graph | grep github.com/gruntwork-io/kubergrunt@v0.6.10
> 
> github.com/gruntwork-io/kubergrunt@v0.6.10 sigs.k8s.io/aws-iam-authenticator@v0.5.1
> github.com/gruntwork-io/kubergrunt@v0.6.10 k8s.io/client-go@v0.18.3
> github.com/gruntwork-io/kubergrunt@v0.6.10 k8s.io/apimachinery@v0.18.3
> github.com/gruntwork-io/kubergrunt@v0.6.10 k8s.io/api@v0.18.3
> github.com/gruntwork-io/kubergrunt@v0.6.10 github.com/urfave/cli@v1.22.4
> github.com/gruntwork-io/kubergrunt@v0.6.10 github.com/stretchr/testify@v1.6.1
> github.com/gruntwork-io/kubergrunt@v0.6.10 github.com/sirupsen/logrus@v1.6.0
> github.com/gruntwork-io/kubergrunt@v0.6.10 github.com/mitchellh/go-homedir@v1.1.0
> github.com/gruntwork-io/kubergrunt@v0.6.10 github.com/gruntwork-io/terratest@v0.30.0
> github.com/gruntwork-io/kubergrunt@v0.6.10 github.com/gruntwork-io/gruntwork-cli@v0.7.0
> github.com/gruntwork-io/kubergrunt@v0.6.10 github.com/blang/semver/v4@v4.0.0
> github.com/gruntwork-io/kubergrunt@v0.6.10 github.com/aws/aws-sdk-go@v1.34.27
> github.com/gruntwork-io/go-commons@v0.8.2 github.com/gruntwork-io/kubergrunt@v0.6.10
> ```
> 
> This line `github.com/gruntwork-io/go-commons@v0.8.2 github.com/gruntwork-io/kubergrunt@v0.6.10` tells that it is required by `go-commons` smile
> 
> So, in summary, `github.com/gruntwork-io/go-commons@v0.8.2` depends on `github.com/gruntwork-io/kubergrunt@v0.6.10` which depends on `github.com/aws/aws-sdk-go@v1.34.27` which then caused this required update smile
> 
> I also found on the [go mod documentation](https://github.com/golang/go/wiki/Modules#why-does-go-mod-tidy-record-indirect-and-test-dependencies-in-my-gomod) that there is the `go mod why` command but it wasn't helpful disappointed

Nice spelunking, and thanks for sharing the results of your investigation!
